### PR TITLE
Move `nn.glob.attention.GlobalAttention` to `nn.aggr.attention.AttentionalAggregation`

### PR DESCRIPTION
**Pull request by Padarn**
Addresses https://github.com/pyg-team/pytorch_geometric/issues/4712

### CODE REVIEW

File: torch_geometric/nn/aggr/attention.py [hunk 1]
```diff
@@ -53,28 +54,21 @@ def reset_parameters(self):
...
-            size (int, optional): The number of graphs in the batch.
-                (default: :obj:`None`)
-        """
-        if batch is None:
-            batch = x.new_zeros(x.size(0), dtype=torch.int64)
+    def forward(self, x: Tensor, index: Optional[Tensor] = None,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
+        self.assert_index_present(index)
```
Comment (rusty1s) [inline torch_geometric/nn/aggr/attention.py:61]: I don't think this is necessarily true. We can replace `scatter_add` with `self.reduce`, and the `softmax` call should support both `index` and `ptr`.

File: torch_geometric/nn/aggr/attention.py [hunk 2]
```diff
@@ -4,13 +4,14 @@
...
 
+from torch_geometric.nn.aggr import Aggregation
 from torch_geometric.utils import softmax
 
 from ..inits import reset
 
 
-class GlobalAttention(torch.nn.Module):
-    r"""Global soft attention layer from the `"Gated Graph Sequence Neural
+class AttentionAggregation(Aggregation):
```
Comment (rusty1s) [inline torch_geometric/nn/aggr/attention.py:13]:
```suggestion
class AttentionalAggregation(Aggregation):
```

File: test/nn/aggr/test_attention.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from torch_geometric.nn.glob import GlobalAttention
+
+
+def test_attention_aggregation():
+    channels, dim_size = (32, 10)
+    gate_nn = Seq(Lin(channels, channels), ReLU(), Lin(channels, 1))
+    nn = Seq(Lin(channels, channels), ReLU(), Lin(channels, channels))
+
+    aggr = AttentionAggregation(gate_nn, nn)
+    assert aggr.__repr__() == (
```
Comment (rusty1s) [inline test/nn/aggr/test_attention.py:16]:
```suggestion
    assert str(aggr) == (
```

File: test/nn/aggr/test_attention.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    x = torch.randn((dim_size**2, channels))
+    index = torch.arange(dim_size, dtype=torch.long)
+    index = index.view(-1, 1).repeat(1, dim_size).view(-1)
+
+    assert aggr(x, index).size() == (dim_size, channels)
+    assert aggr(x, index,
+                dim_size=dim_size + 1).size() == (dim_size + 1, channels)
+
+    # test depreciated
```
Comment (rusty1s) [inline test/nn/aggr/test_attention.py:35]: No reason to test this IMO.

File: test/nn/aggr/test_attention.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        '  (1): ReLU()\n'
+        '  (2): Linear(in_features=32, out_features=32, bias=True)\n'
+        '))')
+
+    x = torch.randn((dim_size**2, channels))
+    index = torch.arange(dim_size, dtype=torch.long)
+    index = index.view(-1, 1).repeat(1, dim_size).view(-1)
+
+    assert aggr(x, index).size() == (dim_size, channels)
+    assert aggr(x, index,
```
Comment (rusty1s) [inline test/nn/aggr/test_attention.py:32]: Test
```python
out = aggr(x, index)
assert out.size() == (dim_size, channels)
torch.allclose(aggr(x, index, dim_size=dim_size), out)
```
instead.

**Review by rusty1s — APPROVED** [review]

**Comment by Padarn** [comment]
Thanks for the comments @rusty1s!

File: test/nn/aggr/test_attention.py (hunk 4 above)
Comment (Padarn) [reply test/nn/aggr/test_attention.py:35]: Agree, added this to do some local tests. Thanks for reminding me

File: torch_geometric/nn/aggr/attention.py [hunk 6]
```diff
@@ -53,28 +54,21 @@ def reset_parameters(self):
...
-                respective graph identifier. (default: :obj:`None`)
-            size (int, optional): The number of graphs in the batch.
-                (default: :obj:`None`)
-        """
-        if batch is None:
-            batch = x.new_zeros(x.size(0), dtype=torch.int64)
+    def forward(self, x: Tensor, index: Optional[Tensor] = None,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
```
[the remark this answers is not in the document]
Comment (Padarn) [reply torch_geometric/nn/aggr/attention.py:60]: We can do this, but right now I think it would be backwards compatibility breaking because of
```
        x = x.unsqueeze(-1) if x.dim() == 1 else x
```
WDYT?

File: test/nn/aggr/test_attention.py (hunk 5 above)
Comment (Padarn) [reply test/nn/aggr/test_attention.py]: I added this, but just to confirm you want to make surer that the output does not change if `dim_size` is provided? Maybe it would be better to test
```
out = aggr(x, index)
assert out.size() == (dim_size, channels)
torch.allclose(aggr(x, index, dim_size=dim_size + 1)[:3], out)
```
WDYT?

File: torch_geometric/nn/aggr/attention.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Padarn) [reply torch_geometric/nn/aggr/attention.py:58]: I've changed for now and removed this line. But still interested in your thought here.

File: torch_geometric/nn/aggr/attention.py (hunk 1 above)
Comment (Padarn) [reply torch_geometric/nn/aggr/attention.py]: you're completely correct, I missed it. Thanks, updated.

File: test/nn/aggr/test_attention.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import torch
+
+from torch_geometric.nn import MLP
+from torch_geometric.nn.aggr import AttentionalAggregation
+
+
+def test_attention_aggregation():
```
Comment (rusty1s) [inline test/nn/aggr/test_attention.py:7]:
```suggestion
def test_attentional_aggregation():
```

File: test/nn/aggr/test_attention.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    channels = 16
+    x = torch.randn(6, channels)
+    index = torch.tensor([0, 0, 1, 1, 1, 2])
+    ptr = torch.tensor([0, 2, 5, 6])
+
+    gate_nn = MLP([channels, 1], act='relu')
+    nn = MLP([channels, channels], act='relu')
+    aggr = AttentionalAggregation(gate_nn, nn)
+    assert str(aggr) == (f'AttentionalAggregation(gate_nn=MLP({channels}, 1),'
+                         f' nn=MLP({channels}, {channels}))')
```
Comment (rusty1s) [inline test/nn/aggr/test_attention.py:17]:
```suggestion
    assert str(aggr) == (f'AttentionalAggregation(gate_nn=MLP({channels}, 1), '
                         f'nn=MLP({channels}, {channels}))')
```

File: torch_geometric/nn/aggr/attention.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/nn/aggr/attention.py:58]: I personally think this is okay. The change on user-side would be quick, and I assume this case is very rare.

**Review by rusty1s — APPROVED** [review]
